File: internal/server/voiceagent/protocol_fixture_test.go
//go:build linux

package voiceagent

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	"github.com/kombifyio/SpeechKit/pkg/speechkit/speaker"
)

// fixturePath is the golden contract file shared with the TypeScript and
// Android clients. The Go structs in protocol.go are the producer of truth;
// this test pins their wire shape to the interchange artifact.
const fixturePath = "voiceagent.v1.json"

func loadVoiceAgentFixture(t *testing.T) map[string]json.RawMessage {
	t.Helper()
	path := filepath.Join("..", "..", "..", "docs", "server", "fixtures", fixturePath)
	raw, err := os.ReadFile(path) // #nosec G304 -- fixed repo-relative test fixture path.
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var doc struct {
		Frames map[string]json.RawMessage `json:"frames"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	if len(doc.Frames) == 0 {
		t.Fatal("fixture has no frames")
	}
	return doc.Frames
}

func fixtureFrame(t *testing.T, frames map[string]json.RawMessage, name string) map[string]any {
	t.Helper()
	entry, ok := frames[name]
	if !ok {
		t.Fatalf("fixture has no frame %q", name)
	}
	var out map[string]any
	if err := json.Unmarshal(entry, &out); err != nil {
		t.Fatalf("parse fixture frame %q: %v", name, err)
	}
	return out
}

// assertVoiceAgentWireEqual marshals v and compares the generic JSON value
// with the golden fixture frame — a byte-order-insensitive producer
// drift-check, matching the dictation stream_protocol_test.go approach.
func assertVoiceAgentWireEqual(t *testing.T, frames map[string]json.RawMessage, name string, v any) {
	t.Helper()
	data, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("marshal %q: %v", name, err)
	}
	var got map[string]any
	if err := json.Unmarshal(data, &got); err != nil {
		t.Fatalf("re-parse %q: %v", name, err)
	}
	want := fixtureFrame(t, frames, name)
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("wire drift on %q:\n got: %s\nwant fixture frame", name, data)
	}
}

func TestVoiceAgentProtocol_GoldenFixtures(t *testing.T) {
	frames := loadVoiceAgentFixture(t)

	// ── client → server ──────────────────────────────────────────────────
	assertVoiceAgentWireEqual(t, frames, "start", StartFrame{
		Type:           MsgStart,
		PersonaID:      "brainstorm",
		SequenceID:     "seq-1",
		Provider:       "deepgram",
		MediaTransport: MediaTransportWebSocket,
		Locale:         "de-DE",
		Thinking:       "low",
	})
	assertVoiceAgentWireEqual(t, frames, "start_full", StartFrame{
		Type:           MsgStart,
		PersonaID:      "brainstorm",
		RoleID:         "moderator",
		SequenceID:     "seq-1",
		Provider:       "gemini",
		MediaTransport: MediaTransportLiveKit,
		Voice:          "Aoede",
		Locale:         "de-DE",
		Model:          "gemini-2.5-flash-native-audio",
		Thinking:       "low",
		ActivityDetection: &ActivityDetectionFrame{
			Automatic:         true,
			StartSensitivity:  "high",
			EndSensitivity:    "low",
			PrefixPaddingMs:   120,
			SilenceDurationMs: 600,
			ActivityHandling:  "start_of_activity_interrupts",
			TurnCoverage:      "turn_includes_only_activity",
		},
		Speaker: &speaker.Options{
			Enabled:           true,
			Diarization:       true,
			ProviderProfileID: "speaker.deepgram.nova-3",
			SpeakersExpected:  2,
		},
		SystemPromptOverride: "Antworte knapp und auf Deutsch.",
	})
	assertVoiceAgentWireEqual(t, frames, "text", TextFrame{
		Type: MsgText,
		Text: "Wie ist das Wetter in Berlin?",
	})
	assertVoiceAgentWireEqual(t, frames, "tool_response", ToolResponseFrame{
		Type:     MsgToolResponse,
		ID:       "t1",
		Name:     "weather",
		Response: map[string]any{"city": "Berlin", "temperature_c": 21.5},
	})
	assertVoiceAgentWireEqual(t, frames, "advance_step", AdvanceStepFrame{
		Type:   MsgAdvanceStep,
		StepID: "step-2",
		Reason: "user_done",
	})
	assertVoiceAgentWireEqual(t, frames, "audio_end", envelope{Type: MsgAudioEnd})
	assertVoiceAgentWireEqual(t, frames, "ping", envelope{Type: MsgPing})
	assertVoiceAgentWireEqual(t, frames, "stop", envelope{Type: MsgStop})
	assertVoiceAgentWireEqual(t, frames, "cancel", envelope{Type: MsgCancel})

	// ── server → client ──────────────────────────────────────────────────
	assertVoiceAgentWireEqual(t, frames, "state_session_ready", StateFrame{
		Type:             MsgState,
		EventFrameFields: EventFrameFields{EventType: EventSessionReady},
		State:            "listening",
		Provider:         "deepgram",
		MediaTransport:   MediaTransportWebSocket,
	})
	assertVoiceAgentWireEqual(t, frames, "state", StateFrame{
		Type:  MsgState,
		State: "speaking",
	})
	assertVoiceAgentWireEqual(t, frames, "input_transcript_partial", TranscriptFrame{
		Type:             MsgInputTranscript,
		EventFrameFields: EventFrameFields{EventType: EventInputPartial},
		Text:             "wie ist das",
		Done:             false,
	})
	assertVoiceAgentWireEqual(t, frames, "input_transcript_final", TranscriptFrame{
		Type:             MsgInputTranscript,
		EventFrameFields: EventFrameFields{EventType: EventInputFinal},
		Text:             "Wie ist das Wetter in Berlin?",
		Done:             true,
	})
	assertVoiceAgentWireEqual(t, frames, "input_transcript_speaker", TranscriptFrame{
		Type:              MsgInputTranscript,
		EventFrameFields:  EventFrameFields{EventType: EventInputFinal},
		Text:              "Wie ist das Wetter in Berlin?",
		Done:              true,
		SpeakerLabel:      "S1",
		PersonID:          "person-7",
		DisplayName:       "Marcel",
		SpeakerConfidence: 0.87,
	})
	assertVoiceAgentWireEqual(t, frames, "output_transcript", TranscriptFrame{
		Type:             MsgOutputTranscript,
		EventFrameFields: EventFrameFields{EventType: EventOutputText},
		Text:             "In Berlin sind es 21 Grad und sonnig.",
		Done:             true,
	})
	assertVoiceAgentWireEqual(t, frames, "tool_call", ToolCallFrame{
		Type:             MsgToolCall,
		EventFrameFields: EventFrameFields{EventType: EventToolCall},
		ID:               "t1",
		Name:             "weather",
		Args:             map[string]any{"city": "Berlin"},
	})
	assertVoiceAgentWireEqual(t, frames, "sequence_step", SequenceStepFrame{
		Type:       MsgSequenceStep,
		SequenceID: "seq-1",
		StepID:     "step-2",
		StepIndex:  2,
		Status:     "entered",
		Reason:     "advance_step",
	})
	assertVoiceAgentWireEqual(t, frames, "event", EventFrame{
		Type:             MsgEvent,
		EventFrameFields: EventFrameFields{EventType: EventTurnEnd},
	})
	assertVoiceAgentWireEqual(t, frames, "interrupted", InterruptedFrame{
		Type:             MsgInterrupted,
		EventFrameFields: EventFrameFields{EventType: EventInterrupted},
	})
	assertVoiceAgentWireEqual(t, frames, "error", ErrorFrame{
		Type:    MsgError,
		Code:    "provider_unavailable",
		Message: `provider "gemini" is not configured on this server`,
	})
	assertVoiceAgentWireEqual(t, frames, "session_end", SessionEndFrame{
		Type:             MsgSessionEnd,
		EventFrameFields: EventFrameFields{EventType: EventSessionEnd},
		Reason:           "idle",
	})
	assertVoiceAgentWireEqual(t, frames, "pong", PongFrame{Type: MsgPong})
}

// TestVoiceAgentFixture_CoversEveryMessageType fails when a message type is
// added to protocol.go without extending the golden fixture (or a fixture
// frame carries a type the protocol does not define).
func TestVoiceAgentFixture_CoversEveryMessageType(t *testing.T) {
	protocolTypes := map[string]bool{
		// client → server
		MsgStart:        true,
		MsgAudioEnd:     true,
		MsgText:         true,
		MsgToolResponse: true,
		MsgPing:         true,
		MsgStop:         true,
		MsgAdvanceStep:  true,
		MsgCancel:       true,
		// server → client
		MsgState:            true,
		MsgInputTranscript:  true,
		MsgOutputTranscript: true,
		MsgToolCall:         true,
		MsgSequenceStep:     true,
		MsgEvent:            true,
		MsgInterrupted:      true,
		MsgError:            true,
		MsgSessionEnd:       true,
		MsgPong:             true,
	}

	frames := loadVoiceAgentFixture(t)
	covered := map[string]bool{}
	for name := range frames {
		frame := fixtureFrame(t, frames, name)
		typ, _ := frame["type"].(string)
		if typ == "" {
			t.Errorf("fixture frame %q has no type field", name)
			continue
		}
		if !protocolTypes[typ] {
			t.Errorf("fixture frame %q carries type %q which protocol.go does not define", name, typ)
			continue
		}
		covered[typ] = true
	}
	for typ := range protocolTypes {
		if !covered[typ] {
			t.Errorf("protocol message type %q has no golden frame in %s", typ, fixturePath)
		}
	}
}

File: clients/typescript/packages/voiceagent-client/src/protocol.fixture.test.ts
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";

import type {
  AdvanceStepFrame,
  ClientFrame,
  ErrorFrame,
  EventFrame,
  SequenceStepFrame,
  ServerFrame,
  SessionEndFrame,
  StartFrame,
  StateFrame,
  TextFrame,
  ToolCallFrame,
  ToolResponseFrame,
  TranscriptFrame,
} from "./protocol.js";

/**
 * Consumer drift-check: replays the golden frames from
 * docs/server/fixtures/voiceagent.v1.json (the interchange artifact produced
 * by internal/server/voiceagent/protocol.go and pinned by its
 * protocol_fixture_test.go) against the protocol.ts frame types.
 */

const fixtureUrl = new URL(
  "../../../../../docs/server/fixtures/voiceagent.v1.json",
  import.meta.url,
);

interface Fixture {
  contract: string;
  frames: Record<string, { type: string } & Record<string, unknown>>;
}

const fixture: Fixture = JSON.parse(readFileSync(fileURLToPath(fixtureUrl), "utf8"));

function frame<T>(name: string): T {
  const entry = fixture.frames[name];
  expect(entry, `fixture frame ${name}`).toBeDefined();
  return entry as T;
}

/** Frame types protocol.ts declares on the ClientFrame union. */
const CLIENT_TYPES = ["start", "text", "tool_response", "advance_step", "audio_end", "ping", "stop"] as const;

/**
 * Client frame types the Go producer defines that protocol.ts does not yet
 * model (tracked drift, see the protocol.go SSOT header): `cancel` is the
 * tap-to-interrupt control frame.
 */
const CLIENT_TYPES_PENDING_TS = ["cancel"] as const;

/** Frame types protocol.ts declares on the ServerFrame union. */
const SERVER_TYPES = [
  "state",
  "input_transcript",
  "output_transcript",
  "tool_call",
  "sequence_step",
  "event",
  "interrupted",
  "error",
  "session_end",
  "pong",
] as const;

describe("voiceagent.v1 fixture", () => {
  it("declares the v1 contract id", () => {
    expect(fixture.contract).toBe("speechkit.voiceagent.v1");
  });

  it("covers every frame type protocol.ts declares (and nothing unknown)", () => {
    const known = new Set<string>([...CLIENT_TYPES, ...CLIENT_TYPES_PENDING_TS, ...SERVER_TYPES]);
    const seen = new Set(Object.values(fixture.frames).map((f) => f.type));
    for (const entry of Object.values(fixture.frames)) {
      expect(known, `unknown frame type ${entry.type}`).toContain(entry.type);
    }
    for (const type of known) {
      expect(seen, `no golden frame for type ${type}`).toContain(type);
    }
  });

  it("start frames decode as StartFrame", () => {
    const start = frame<StartFrame>("start");
    expect(start.type).toBe("start");
    expect(start.persona_id).toBe("brainstorm");
    expect(start.provider).toBe("deepgram");
    expect(start.media_transport).toBe("websocket");
    expect(start.thinking).toBe("low");

    const full = frame<StartFrame>("start_full");
    expect(full.media_transport).toBe("livekit");
    expect(full.activity_detection?.automatic).toBe(true);
    expect(full.activity_detection?.silence_duration_ms).toBe(600);
    expect(full.speaker?.providerProfileId).toBe("speaker.deepgram.nova-3");
    expect(full.system_prompt_override).toBeTruthy();
  });

  it("client frames decode into the ClientFrame union", () => {
    const text = frame<TextFrame>("text");
    expect(text.text).toBe("Wie ist das Wetter in Berlin?");

    const toolResponse = frame<ToolResponseFrame>("tool_response");
    expect(toolResponse.id).toBe("t1");
    expect(toolResponse.name).toBe("weather");
    expect(toolResponse.response).toEqual({ city: "Berlin", temperature_c: 21.5 });

    const advance = frame<AdvanceStepFrame>("advance_step");
    expect(advance.step_id).toBe("step-2");

    for (const name of ["audio_end", "ping", "stop"] as const) {
      const control = frame<ClientFrame>(name);
      expect(control.type).toBe(name);
      expect(Object.keys(control)).toEqual(["type"]);
    }
  });

  it("state frames decode as StateFrame", () => {
    const ready = frame<StateFrame>("state_session_ready");
    expect(ready.state).toBe("listening");
    expect(ready.event_type).toBe("session_ready");

    const speaking = frame<StateFrame>("state");
    expect(speaking.state).toBe("speaking");
  });

  it("transcript frames decode for both directions", () => {
    const partial = frame<TranscriptFrame>("input_transcript_partial");
    expect(partial.type).toBe("input_transcript");
    expect(partial.done).toBe(false);
    expect(partial.text).toBe("wie ist das");

    const final = frame<TranscriptFrame>("input_transcript_final");
    expect(final.done).toBe(true);

    // Speaker-attribution fields are additive server-side extras protocol.ts
    // does not yet surface; the frame must still narrow as TranscriptFrame.
    const attributed = frame<TranscriptFrame>("input_transcript_speaker");
    expect(attributed.done).toBe(true);

    const output = frame<TranscriptFrame>("output_transcript");
    expect(output.type).toBe("output_transcript");
    expect(output.text).toBe("In Berlin sind es 21 Grad und sonnig.");
  });

  it("tool_call, sequence_step, event, interrupted decode", () => {
    const toolCall = frame<ToolCallFrame>("tool_call");
    expect(toolCall.id).toBe("t1");
    expect(toolCall.args).toEqual({ city: "Berlin" });

    const step = frame<SequenceStepFrame>("sequence_step");
    expect(step.step_id).toBe("step-2");
    expect(step.step_index).toBe(2);
    expect(step.status).toBe("entered");

    const event = frame<EventFrame>("event");
    expect(event.event_type).toBe("turn_end");

    const interrupted = frame<ServerFrame>("interrupted");
    expect(interrupted.type).toBe("interrupted");
  });

  it("error, session_end, pong decode", () => {
    const error = frame<ErrorFrame>("error");
    expect(error.code).toBe("provider_unavailable");
    expect(error.message).toBeTruthy();

    const end = frame<SessionEndFrame>("session_end");
    expect(end.reason).toBe("idle");

    const pong = frame<ServerFrame>("pong");
    expect(pong.type).toBe("pong");
  });
});
